### DISCUSSION TITLE
Podman + Quadlet on Pi Zero 2 W

### DISCUSSION THREAD

**Discussion by torntrousers** [containers/podman#21396] [Q&A]
Hello! I'd really like to be able to use Podman and Quadlet on a Raspberry Pi Zero 2 W, can you help me get it working?

Suggestion from [here ](https://github.com/containers/podman.io_old/issues/383#issuecomment-1913559904) is to start this discussion.

Any pointers? 

Is the only way to build from source, and if so any tips on doing that? I've started following the instructions [here](https://podman.io/docs/installation#building-from-source) and get to "Make sure that the Linux kernel supports user namespaces...If not, please update the kernel" and it looks like it does not (on the Pi Zero which is running Debian bullseye), and I'm not sure how to update the kernel.

Can i do this build not on the Pi Zero, on my x86 windows machine in WSL perhaps?

**Comment by afbjorklund** [comment]
~~On debian, you need to enable `kernel.unprivileged_userns_clone` to enable user namespaces.~~ Maybe some other tweaks, too...

Maybe not with the newer kernel though. Rootless containers seem to be working without it, at least with `nerdctl`

`Linux raspberrypi 6.1.0-rpi7-rpi-v8 #1 SMP PREEMPT Debian 1:6.1.63-1+rpt1 (2023-11-24) aarch64 GNU/Linux`

But you can build on your other machine, using `GOOS=linux GOARCH=arm64` cross-compile. Probably with `CGO_ENABLED=0` (static)

----

EDIT: Kernel settings were only needed for Debian 10, and before.

Things have changed a bit, since Raspberry Pi 2 and HypriotOS :-)

https://blog.hypriot.com/getting-started-with-docker-on-your-arm-device/

A similarly preconfigured SD card for Raspberry Pi 3+ would be nice.

**Reply by afbjorklund** [reply]
Once upon a time there was a project for running Docker on your Raspberry Pi, called https://hypriot.com/

Maybe something similar could be needed for Podman, but most users would just run Fedora in that case?

I was using it for the Lima physical machines, and I suppose one _could_ add support for the legacy platforms

* https://github.com/lima-vm/lima/discussions/2029

* Raspberry Pi OS Lite (64-bit)

* Debian 12 Bookworm (stable)

```console
anders@raspberrypi:~ $ nerdctl version
Client:
 Version:	v1.7.2
 OS/Arch:	linux/arm64
 Git commit:	e32c4b023bf41e5c8325cfb893a53cefb5fc68ed
 buildctl:
  Version:	v0.12.4
  GitCommit:	833949d0f7908608b00ab6b93b8f92bdb147fcca

Server:
 containerd:
  Version:	v1.7.11
  GitCommit:	64b8a811b07ba6288238eefc14d898ee0b5b99ba
 runc:
  Version:	1.1.10
  GitCommit:	v1.1.10-0-g18a0cb0f

```

**Reply by siretart** [reply]
>On debian, you need to enable kernel.unprivileged_userns_clone to enable user namespaces. Maybe some other tweaks, too...

that used to be the case for Debian 10. Starting with Debian 11 (released August 2021), podman runs out of the box.

**Reply by afbjorklund** [reply]
Right, the codenames aren't mapping correctly from swap memory. Wonder why it failed above, then.

**Reply by afbjorklund** [reply]
If the rainy weather comes back, I will try to install Podman again. Hope I don't need to install a Rust compiler, too.

**Reply by siretart** [reply]
no, you can just install the precompiled binaries using `apt install podman`

**Reply by afbjorklund** [reply]
You mean the `sid` ones ? Or recompile the source packages, perhaps

**Reply by afbjorklund** [reply]
The old version did install without issues, for what it is worth.

```console
anders@raspberrypi:~ $ podman version
Client:       Podman Engine
Version:      4.3.1
API Version:  4.3.1
Go Version:   go1.19.8
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/arm64
anders@raspberrypi:~ $ apt list podman
Listing... Done
podman/stable,now 4.3.1+ds1-8+b1 arm64 [installed]
podman/stable 4.3.1+ds1-8+b1 armhf
```

But I haven't tested quadlet or farm, or any new stuff.

And I still got `vfs` set up out of the box, unfortunately.

**Reply by torntrousers** [reply]
```apt install podman``` give version 3.0.1 though which is pretty old and I want Quadlet so need at least 4.4, right?

**Reply by afbjorklund** [reply]
Right, with bullseye (oldstable) you get 3.0.1. Probably should have podman.io link to https://packages.debian.org/podman

**Reply by afbjorklund** [reply]
Could be worse, with CentOS XP you get 1.6.4

**Reply by afbjorklund** [reply]
I think I will just have people get another SD card

**Reply by afbjorklund** [reply]
But according to https://fedoraproject.org/wiki/Architectures/ARM/Raspberry_Pi it is not supported yet (Zero 2)

> Official support for the Zero 2W is under review.

> WiFi on the Raspberry Pi Zero 2W and Raspberry Pi 400 is currently unsupported.

I guess you are _supposed_ to be using something immutable and closer to Fedora CoreOS, like [Fedora IoT](https://www.redhat.com/sysadmin/fedora-iot-raspberry-pi) ?

**Comment by siretart** [comment]
Please consider running debian testing, see https://wiki.debian.org/DebianTesting for details.
There are possible compromises that are unacceptable to you, like rate of updates, the associated bandwith costs, etc. If that is the case, consider installing the packages from debian/testing on your bookworm systems using Apt Pinning: https://wiki.debian.org/AptConfiguration?action=show&redirect=AptPreferences

**Reply by afbjorklund** [reply]
Most users are looking for a custom repo, with **only** the updated packages (for podman, and dependencies)...

Or something different, like OBS with Nix packages (like cri-o) or perhaps a "podman-full" tarball (like nerdctl)?

**Reply by siretart** [reply]
> Most users are looking for a custom repo.

Maintaining such a repo is a huge effort. I've been approached on several occasions to provide updated packages in bookworm-backports, but I don't currently find the time for that.

At this point, your best options are:
- use debian/testing
- use packages from debian/testing on earlier versions of debian (or ubuntu, I guess)
- compile yourself

**Reply by afbjorklund** [reply]
* use fedora

**Comment by torntrousers** [comment]
Ok so trying in Windows WSL running Ubuntu, the prereqs installed ok, and "zgrep CONFIG_USER_NS /proc/config.gz" shows "CONFIG_USER_NS=y" but "sudo sysctl kernel.unprivileged_userns_clone=1" gives "No such file or directory" and "sudo echo 'kernel.unprivileged_userns_clone=1' > /etc/sysctl.d/userns.conf" gives "/etc/sysctl.d/userns.conf: Permission denied"

**Reply by siretart** [reply]
The WSL kernel is working fine. I'm semi-regularity running debian/testing in WSL2, and can confirm podman runs just fine there. On WSL running Ubuntu, podman should run fine as well, it's the same kernel after all.

**Reply by torntrousers** [reply]
I want to run podman on the Pi though so just trying to do the build of Podman for that in Ubuntu

**Reply by siretart** [reply]
oh, I didn't realize that WSL2 is also available on amd64? -- amd64 binaries won't work on rasberry pis, so unless you do quite complicated cross-compiling, you might be better off trying this build:
https://ftp.debian.org/debian/pool/main/libp/libpod/podman_4.9.0+ds1-2_arm64.deb

**Reply by torntrousers** [reply]
that sounds promising, but how - 
```
ant@pizero2:~/podman $ ls -al pod*
-rwxrwxrwx 1 ant ant 10529760 Jan 25 13:01 podman_4.9.0+ds1-2_arm64.deb
ant@pizero2:~/podman $ sudo apt install podman_4.9.0+ds1-2_arm64.deb
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package podman_4.9.0+ds1-2_arm64.deb
```

**Reply by torntrousers** [reply]
oh, with dpkg, ...
```
sudo dpkg -i podman_4.9.0+ds1-2_arm64.deb
(Reading database ... 54518 files and directories currently installed.)
Preparing to unpack podman_4.9.0+ds1-2_arm64.deb ...
Unpacking podman (4.9.0+ds1-2) over (3.0.1+dfsg1-3+deb11u4) ...
dpkg: dependency problems prevent configuration of podman:
 podman depends on libc6 (>= 2.34); however:
  Version of libc6:arm64 on system is 2.31-13+rpt2+rpi1+deb11u7.
 podman depends on libsqlite3-0 (>= 3.36.0); however:
  Version of libsqlite3-0:arm64 on system is 3.34.1-3.
 podman depends on libsubid4 (>= 1:4.11.1); however:
  Package libsubid4 is not installed.

dpkg: error processing package podman (--install):
 dependency problems - leaving unconfigured
Processing triggers for man-db (2.9.4-2) ...
Errors were encountered while processing:
 podman
```

**Reply by torntrousers** [reply]
```
sudo apt install libsubid4
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Package libsubid4 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'libsubid4' has no installation candidate
```

**Reply by torntrousers** [reply]
trying the 4.8.3 release - 
```
sudo dpkg -i podman-docker_4.8.3+ds1-2_arm64.deb
Selecting previously unselected package podman-docker.
dpkg: regarding podman-docker_4.8.3+ds1-2_arm64.deb containing podman-docker:
 podman-docker conflicts with docker.io
  docker.io (version 20.10.5+dfsg1-1+deb11u2) is present and installed.

dpkg: error processing archive podman-docker_4.8.3+ds1-2_arm64.deb (--install):
 conflicting packages - not installing podman-docker
Errors were encountered while processing:
 podman-docker_4.8.3+ds1-2_arm64.deb
```

**Reply by torntrousers** [reply]
woot!
```
ant@pizero2:~ $ podman --version
podman version 4.9.0
```

**Reply by siretart** [reply]
I hope you understand now why I suggested to look into "apt pinning", or if that is too much hassle, just run debian/testing in the first place.

there are quite a number of dependencies and package relationships to keep track of (and thus, to backport, which is what you were originally asking for)

**Comment by torntrousers** [comment]
> woot!
> 
> ```
> ant@pizero2:~ $ podman --version
> podman version 4.9.0
> ```

Steps were:
1) to get Debian 12 instead off 11 which seems to be the default:
- in Pi Imager tool, in the Device Type select "No Filtering" (otherwise Bookworm doesn't show as a choice)
- For Choose OS select Raspberry Pi OS 64bit Debian Bookworm
2) Boot up the Pi and then:
- ```cat /etc/os-release``` should show Debian 12
- ```sudo apt-get update && sudo apt-get upgrade```
- ```wget https://ftp.debian.org/debian/pool/main/libp/libpod/podman_4.9.0+ds1-2_arm64.deb```
- ```sudo dpkg -i podman_4.9.0+ds1-2_arm64.deb``` (which fails complaining about libsubid4 missing
- ```sudo apt install libsubid4``` (which fails complaining about a bunch of things)
- ```sudo apt --fix-broken install```
- ```sudo apt-get install uidmap```
- ```sudo apt-get install slirp4netns``` 
- ```sudo apt-get install crun```


(if it hangs reading changelogs update the swap size: https://forums.raspberrypi.com/viewtopic.php?t=360234#p2160951)

and now it should all work:
```
podman --version
podman version 4.9.0
```
```
podman run --rm docker.io/library/hello-world
```

Thanks for all the help @afbjorklund and @siretart !

**Reply by afbjorklund** [reply]
We still need a repository, to handle the updates (4.9.1, etc) and the missing dependencies.

**Reply by afbjorklund** [reply]
Note that you can run podman on the default 32-bit OS (armv7l) too, if you want to / have to.

**Reply by torntrousers** [reply]
also needs uidmap and slirp4netns. have updated the steps in https://github.com/containers/podman/discussions/21396#discussioncomment-8271980

**Comment by torntrousers** [comment]
Anyway, this is all working and i have a running Quadlet but one final question - I see in the log there is a warning. I do ```systemctl --user status pi-blink.service``` and in the output it shows: 
```Feb 23 05:55:05 pizeroups conmon[749]: conmon 523eb0be7e54b7e0bfcb <nwarn>: Failed to add inotify watch for /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/pi-blink.service/libpod-payload-523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/memory.events```

What does that mean and is it anything to worry about?

Here is the full output:
```
● pi-blink.service
     Loaded: loaded (/home/ant/.config/containers/systemd/pi-blink.container; generated)
     Active: active (running) since Fri 2024-02-23 05:55:05 GMT; 40s ago
   Main PID: 749 (conmon)
      Tasks: 3 (limit: 176)
        CPU: 1.305s
     CGroup: /user.slice/user-1000.slice/user@1000.service/app.slice/pi-blink.service
             ├─libpod-payload-523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2
             │ └─754 python3 blink.py
             └─runtime
               ├─734 /usr/bin/slirp4netns --disable-host-loopback --mtu=65520 --enable-sandbox --enable-seccomp --enable-ipv6 -c -r 3 -e 4 --netns-type=path /run/user/1000/netns/netns-b7600762-499e-3d8a-90ba-c07eadd7e314 tap0
               └─749 /usr/bin/conmon --api-version 1 -c 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2 -u 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2 -r /usr/bin/crun -b /home/ant/.local/share/containers/storage/overlay-containers/523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/userdata -p /run/user/1000/containers/overlay-containers/523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/userdata/pidfile -n systemd-pi-blink --exit-dir /run/user/1000/libpod/tmp/exits --full-attach -l journald --log-level warning --syslog --runtime-arg --log-format=json --runtime-arg --log --runtime-arg=/run/user/1000/containers/overlay-containers/523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/userdata/oci-log --conmon-pidfile /run/user/1000/containers/overlay-containers/523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/userdata/conmon.pid --exit-command /usr/bin/podman --exit-command-arg --root --exit-command-arg /home/ant/.local/share/containers/storage --exit-command-arg --runroot --exit-command-arg /run/user/1000/containers --exit-command-arg --log-level --exit-command-arg warning --exit-command-arg --cgroup-manager --exit-command-arg systemd --exit-command-arg --tmpdir --exit-command-arg /run/user/1000/libpod/tmp --exit-command-arg --network-config-dir --exit-command-arg "" --exit-command-arg --network-backend --exit-command-arg netavark --exit-command-arg --volumepath --exit-command-arg /home/ant/.local/share/containers/storage/volumes --exit-command-arg --db-backend --exit-command-arg sqlite --exit-command-arg --transient-store=false --exit-command-arg --runtime --exit-command-arg crun --exit-command-arg --storage-driver --exit-command-arg overlay --exit-command-arg --events-backend --exit-command-arg journald --exit-command-arg container --exit-command-arg cleanup --exit-command-arg --rm --exit-command-arg 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2

Feb 23 05:55:03 pizeroups systemd[575]: Starting pi-blink.service...
Feb 23 05:55:04 pizeroups podman[651]: 2024-02-23 05:55:04.645906662 +0000 GMT m=+0.799316250 system refresh
Feb 23 05:55:04 pizeroups podman[651]: 2024-02-23 05:55:04.700712495 +0000 GMT m=+0.854122969 image pull 67d06bd996ddc11bf7253f1a583fcacec3d950ad695ff600f628f1d40b445d47 docker.io/quarklinkdev/pi-blink:latest
Feb 23 05:55:04 pizeroups podman[651]: 2024-02-23 05:55:04.878003589 +0000 GMT m=+1.031413125 container create 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2 (image=docker.io/quarklinkdev/pi-blink:latest, name=systemd-pi-blink, PODMAN_SYSTEMD_UNIT=pi-blink.service)
Feb 23 05:55:05 pizeroups conmon[749]: conmon 523eb0be7e54b7e0bfcb <nwarn>: Failed to add inotify watch for /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/pi-blink.service/libpod-payload-523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2/memory.events
Feb 23 05:55:05 pizeroups podman[651]: 2024-02-23 05:55:05.30912661 +0000 GMT m=+1.462536198 container init 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2 (image=docker.io/quarklinkdev/pi-blink:latest, name=systemd-pi-blink, PODMAN_SYSTEMD_UNIT=pi-blink.service)
Feb 23 05:55:05 pizeroups podman[651]: 2024-02-23 05:55:05.333983641 +0000 GMT m=+1.487393177 container start 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2 (image=docker.io/quarklinkdev/pi-blink:latest, name=systemd-pi-blink, PODMAN_SYSTEMD_UNIT=pi-blink.service)
Feb 23 05:55:05 pizeroups systemd[575]: Started pi-blink.service.
Feb 23 05:55:05 pizeroups pi-blink[651]: 523eb0be7e54b7e0bfcb58d16a74bc8f35f0c0cd58b2be09e763ce662e9609b2
Feb 23 05:55:06 pizeroups systemd-pi-blink[749]: Blinking: Colour.RGB, speed=1.0, and pi5=False
```
